src/live-programming.rst

```diff
@@ -47,3 +47,210 @@ TODO: in music
+
+.. container:: bib-item
+
+  .. bibliography:: programmable_PE2.bib
+    :filter: key == "saraswat1994factory"
+
+.. container:: bib-item
+
+  .. bibliography:: programmable_PE3.bib
+    :filter: key == "nitzan1976industrial"
+
+As the development of technology, industrial automation enables more flexible work, which enables the same automated equipment to be programmed to perform different tasks. The development of the controller systems gave the growth of research on designing corresponding interface for the controller systems. Here are some papers on different interface architecture for programmable controller systems.
+
+For now, there are many researches on programmable logic circuit. For college learning and industrial production, software for programming logic circuits is widely used.
+
+However, it is quite hard to divide these application as a branch of programmable physical environment. Most of the applications in factory manufacture are surely including programmable physical things, but there's merely elements of an "environment". The interaction between people and the systems is still limited, especially for the sensing part. The liveness of such systems may be still at a low level, as in industrial automation, it is hard to rapidly see the results from the changes to the programs.
+
+Internet of things (IOT)
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+A possible future is that computing will be ouside the realm of the traditional desktops. Many of the objects surrouding people's life will be on the network in certain form. There are more ways for people to have interaction with such a network of physical things.
+
+.. container:: bib-item
+
+  .. bibliography:: programmable_PE2.bib
+    :filter: key == "lee2015iot"
+
+.. container:: bib-item
+
+  .. bibliography:: programmable_PE3.bib
+    :filter: key == "luigi2012socialiot"
+
+An example of the interactive interface to IOT is the Echo Dot of Amazon, a kind of intelligent virtual assistants. Similar products in people's life include Siri, Assistant Cortana. Such intelligent virtual assistants can provide many functions for people's daily life. With connection to cloud database, they can compute various requests from people. Besides fetching data from cloud database, they can have control of other physical objects in the network. They serve as an interface to IOT and internet for people to interact with.
+
+However, though the intelligent virtual assistants give a good start to programmable IOT interfaces, there are worries about the security and privacy. There's still a long way for it to become a mature programmable physical environment.
```

Comment:
You should mention that these assistants are not very programmable. They respond to immediate commands but it is difficult to give them more complex instructions.